### PR TITLE
Add PHPStan check

### PR DESCRIPTION
**Pull request by franmomu**

### CODE REVIEW

File: tests/Model/ModelManagerTest.php [hunk 1]
```diff
@@ -13,16 +13,343 @@
...
+        $this->assertSame($metadata->fieldMappings[$lastPropertyName]['type'], 'string');
+
+        [$metadata, $lastPropertyName] = $modelManager
+            ->getParentMetadataForProperty($containerDocumentClass, 'embeddedDocument.plainField');
+        $this->assertSame($metadata->fieldMappings[$lastPropertyName]['type'], 'boolean');
+
+        $this->assertSame($metadata->fieldMappings[$lastPropertyName]['type'], 'boolean');
+    }
+
+    public function getMetadataForEmbeddedDocument()
```
Comment (phansys) [inline tests/Model/ModelManagerTest.php]: Could we add return declarations for these methods?

File: .github/workflows/qa.yml [hunk 2]
```diff
@@ -0,0 +1,12 @@
+on: [push, pull_request]
+name: Quality assurance
+jobs:
+    phpstan:
+        name: PHPStan
+        runs-on: ubuntu-latest
+        steps:
+            - uses: actions/checkout@master
```
Comment (OskarStark) [inline .github/workflows/qa.yml]: We should use @v1 everywhere

File: .github/workflows/qa.yml [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+            - name: PHPStan
+              uses: "docker://oskarstark/phpstan-ga"
```
Comment (OskarStark) [inline .github/workflows/qa.yml]: @0.12.2 ?

Comment (core23) [reply .github/workflows/qa.yml]: Is there a reason, why we shouldn't use the latest version? We're doing the same for every other QA tool (php-cs-fixer, phpunit, ...)

Comment (OskarStark) [reply .github/workflows/qa.yml]: Ok in this case

Comment (franmomu) [reply .github/workflows/qa.yml]: What about the PHP extension mongodb? can I do something?

File: src/Filter/AbstractDateFilter.php [hunk 4]
```diff
@@ -49,34 +49,44 @@ public function filter(ProxyQueryInterface $queryBuilder, $alias, $field, $data)
         $data['type'] = !isset($data['type']) || !is_numeric($data['type']) ? DateType::TYPE_EQUAL : $data['type'];
 
         // Some types do not require a value to be set (NULL, NOT NULL).
-        if (!$this->typeRequiresValue($data['type']) && !$data['value']) {
+        if (!$this->typeRequiresValue($data['type']) && (!\array_key_exists('value', $data) || !$data['value'])) {
```
Comment (phansys) [inline src/Filter/AbstractDateFilter.php]:
```suggestion
        if (!($data['value'] ?? false) && !$this->typeRequiresValue($data['type'])) {
```

File: src/Model/ModelManager.php [hunk 5]
```diff
@@ -359,6 +359,17 @@ public function getExportFields($class)
      */
     public function getModelInstance($class)
     {
+        $r = new \ReflectionClass($class);
+        if ($r->isAbstract()) {
+            throw new \RuntimeException(sprintf('Cannot initialize abstract class: %s', $class));
```
Comment (phansys) [inline src/Model/ModelManager.php]:
```suggestion
            throw new \InvalidArgumentException(sprintf('Cannot initialize abstract class: %s', $class));
```

File: src/Model/ModelManager.php [hunk 6]
```diff
@@ -443,7 +454,7 @@ public function modelReverseTransform($class, array $array = [])
 
             if ($reflClass->hasMethod($setter)) {
                 if (!$reflClass->getMethod($setter)->isPublic()) {
-                    throw new PropertyAccessDeniedException(sprintf('Method "%s()" is not public in class "%s"', $setter, $reflClass->getName()));
+                    throw new \RuntimeException(sprintf('Method "%s()" is not public in class "%s"', $setter, $reflClass->getName()));
```
Comment (phansys) [inline src/Model/ModelManager.php]: IMO, `BadMethodCallException` could be a better choice here; but `DomainException` should work for both (the method and the property).

**Comment by greg0ire** [comment]
@franmomu can you please extract your 2 first commits into a separate PR? I feel that the issue you have will not be easy to fix… a solution could be to run Composer with `--ignore-platform-reqs`, or to install the mongodb extension on @OskarStark 's Docker image, but both solutions sound like they would have unwanted consequences (installing packages that are not compatible with the targeted version of php, or ending up with an image with many, many, hopefully compatible extensions)

**Comment by franmomu** [comment]
> @franmomu can you please extract your 2 first commits into a separate PR? I feel that the issue you have will not be easy to fix… a solution could be to run Composer with `--ignore-platform-reqs`, or to install the mongodb extension on @OskarStark 's Docker image, but both solutions sound like they would have unwanted consequences (installing packages that are not compatible with the targeted version of php, or ending up with an image with many, many, hopefully compatible extensions)

Sure! I forgot about this one, I remember to check this and try to see a way to enable extension, but I didn't find an easy way.

**Comment by greg0ire** [comment]
@franmomu another option might be using stubs: https://github.com/doctrine/dbal/pull/3951#discussion_r408377525

**Comment by franmomu** [comment]
> @franmomu another option might be using stubs: [doctrine/dbal#3951 (comment)](https://github.com/doctrine/dbal/pull/3951#discussion_r408377525)

[I've just tried](https://github.com/franmomu/SonataDoctrineMongoDBAdminBundle/runs/587128622?check_suite_focus=true#step:4:8), but the problem is that `composer install` fails on install 😞

**Comment by greg0ire** [comment]
Ah that's right, I forgot. I think it should be combined with `--ignore-platform-reqs`, which will imply contributing something to the phpstan-ga repo

**Comment by franmomu** [comment]
I'll make a PR soon like the one for [the dev dependencies](https://github.com/OskarStark/phpstan-ga/pull/22)

**Pull request closed without merging**